core/markets/storage/provider/impl/provider_impl.cpp

```diff
@@ -346,26 +330,43 @@ namespace fc::markets::storage::provider {
     return std::move(cid);
   }
 
-  void StorageProviderImpl::sendSignedResponse(
+  outcome::result<void> StorageProviderImpl::sendSignedResponse(
       std::shared_ptr<MinerDeal> deal) {
+    OUTCOME_TRY(chain_head, api_->ChainHead());
+    OUTCOME_TRY(tipset_key, chain_head.makeKey());
+    OUTCOME_TRY(worker_info,
+                api_->StateMinerInfo(
+                    deal->client_deal_proposal.proposal.provider, tipset_key));
+    OUTCOME_TRY(worker_key_address,
+                api_->StateAccountKey(worker_info.worker, tipset_key));
     Response response{.state = deal->state,
                       .message = deal->message,
                       .proposal = deal->proposal_cid,
                       // if deal is not published, set any valid value
                       .publish_message = deal->publish_cid
                                              ? deal->publish_cid.get()
                                              : deal->proposal_cid};
-    // TODO sign response
-    SignedResponse signed_response{.response = response};
-    // TODO handle if stream is absent
-    auto stream = connections_[deal->proposal_cid];
-    stream->write(
-        signed_response,
-        [self{shared_from_this()}, stream, deal](
-            outcome::result<size_t> maybe_res) {
-          SELF_FSM_HALT_ON_ERROR(maybe_res, "Write deal response error", deal);
-          SELF_FSM_SEND(deal, ProviderEvent::ProviderEventDealPublished);
-        });
+    OUTCOME_TRY(encoded_response, codec::cbor::encode(response));
+    OUTCOME_TRY(signature,
+                api_->WalletSign(worker_key_address, encoded_response));
+    SignedResponse signed_response{.response = response,
+                                   .signature = signature};
+    OUTCOME_TRY(stream, getStream(deal->proposal_cid));
+    stream->write(signed_response,
+                  [self{shared_from_this()}, stream, deal](
+                      outcome::result<size_t> maybe_res) {
+                    if (maybe_res.has_error()) {
+                      // assume client disconnected
+                      self->logger_->error("Write deal response error. "
+                                           + maybe_res.error().message());
+                      return;
```

Comment:
should we update the deal state before returning?

Reply:
It is updated in CALLBACK_ACTION.